src/InterchainCreate2FactoryRouter.sol

```diff
@@ -0,0 +1,280 @@
+     * @param _hookMetadata The hook metadata to override with for the hook set by the owner
+     */
+    function deployContract(
+        uint32 _destination,
+        bytes32 _ism,
+        bytes32 _salt,
+        bytes memory _bytecode,
+        bytes memory _hookMetadata
+    )
+        external
+        payable
+        returns (bytes32)
+    {
+        bytes memory _body = InterchainCreate2FactoryMessage.encode(msg.sender, _ism, _salt, _bytecode, new bytes(0));
+
+        return _dispatchMessage(_destination, _ism, _body, _hookMetadata);
+    }
+
+    /**
+     * @notice Deploys and initialize a contract on the `_destination` chain
+     * @param _destination The remote domain
+     * @param _ism The address of the remote ISM, zero address for using the remote mailbox default ISM
+     * @param _salt The salt used for deploying the contract with CREATE2
+     * @param _bytecode The bytecode of the contract to deploy
+     * @param _initCode The initialization that is called after the contract is deployed
+     */
+    function deployContractAndInit(
+        uint32 _destination,
+        bytes32 _ism,
+        bytes32 _salt,
+        bytes memory _bytecode,
+        bytes memory _initCode
+    )
+        external
+        payable
+        returns (bytes32)
+    {
+        bytes memory _body = InterchainCreate2FactoryMessage.encode(msg.sender, _ism, _salt, _bytecode, _initCode);
+
+        return _dispatchMessage(_destination, _ism, _body, "");
+    }
+
+    /**
+     * @notice Deploys and initialize a contract on the `_destination` chain with hook metadata
+     * @param _destination The remote domain
+     * @param _ism The address of the remote ISM, zero address for using the remote mailbox default ISM
+     * @param _salt The salt used for deploying the contract with CREATE2
+     * @param _bytecode The bytecode of the contract to deploy
+     * @param _initCode The initialization that is called after the contract is deployed
+     * @param _hookMetadata The hook metadata to override with for the hook set by the owner
+     */
+    function deployContractAndInit(
+        uint32 _destination,
+        bytes32 _ism,
+        bytes32 _salt,
+        bytes memory _bytecode,
+        bytes memory _initCode,
+        bytes memory _hookMetadata
+    )
+        external
+        payable
+        returns (bytes32)
+    {
+        bytes memory _body = InterchainCreate2FactoryMessage.encode(msg.sender, _ism, _salt, _bytecode, _initCode);
+
+        return _dispatchMessage(_destination, _ism, _body, _hookMetadata);
+    }
+
+    /**
+     * @notice Returns the gas payment required to dispatch a given messageBody to the given domain's router with gas
+     * limit override.
+     * @param _destination The domain of the destination router.
+     * @param _messageBody The message body to be dispatched.
+     * @param _hookMetadata The hook metadata to override with for the hook set by the owner
+     */
+    function quoteGasPayment(
```

Comment:
I would be important to offer to override the hook itself as well (if the caller wants to pay a different relayer, or use a different ISM that needs a post dispatch hook (like rollup bridges)

Reply:
if we offer to override the hook here then we should do the same for `deployContract`, right?
Something like:
```solidity
function quoteDispatch(
        uint32 _destinationDomain,
        bytes32 _router,
        bytes memory _messageBody,
        bytes memory _hookMetadata,
        address _hook
    ) internal view returns (uint256) {
        return
            mailbox.quoteDispatch(
                _destinationDomain,
                _router,
                _messageBody,
                _hookMetadata,
                IPostDispatchHook(_hook)
            );
    }
```

Reply:
Indeed